File: Linked List/circular/traversing/list.java
// Circular Singly Linked List - Traversing (Java)
// Traverse once around the circle starting from head.

class Node {
    int data;
    Node next;
    Node(int value) { this.data = value; this.next = null; }
}

public class list {
    static void traverse(Node head) {
        if (head == null) { System.out.println(); return; }
        Node current = head;
        do {
            System.out.print(current.data + " ");
            current = current.next;
        } while (current != head);
        System.out.println();
    }

    public static void main(String[] args) {
        Node n1 = new Node(10), n2 = new Node(20), n3 = new Node(30);
        n1.next = n2; n2.next = n3; n3.next = n1;
        Node head = n1;
        System.out.print("Circular traverse: ");
        traverse(head);
    }
}

File: Linked List/circular/traversing/list.py
# Circular Singly Linked List - Traversing (Python)
# Traverse once around the circle starting from head.

class Node:
    def __init__(self, value: int) -> None:
        self.data = value
        self.next: "Node | None" = None


def traverse(head: "Node | None") -> None:
    if head is None:
        print()
        return
    current = head
    while True:
        print(current.data, end=" ")
        current = current.next
        if current is head:
            break
    print()


if __name__ == "__main__":
    n1 = Node(10); n2 = Node(20); n3 = Node(30)
    n1.next = n2; n2.next = n3; n3.next = n1
    head = n1
    print("Circular traverse:", end=" ")
    traverse(head)


